pkg/controller/core/clusterqueue_controller_test.go

```diff
@@ -489,3 +487,82 @@ func TestRecordResourceMetrics(t *testing.T) {
 		})
 	}
 }
+
+func TestStartSnapshot(t *testing.T) {
+	cqName := "test-cq"
+	lqName := "test-lq"
+	const lowPrio, highPrio = 0, 100
+	defaultWls := &kueue.WorkloadList{
+		Items: []kueue.Workload{
+			*utiltesting.MakeWorkload("one", "").Queue(lqName).Priority(highPrio).Obj(),
+			*utiltesting.MakeWorkload("two", "").Queue(lqName).Priority(lowPrio).Obj(),
+		},
+	}
+	testCases := map[string]struct {
+		queueVisibilityUpdateInterval        time.Duration
+		queueVisibilityClusterQueuesMaxCount int32
+		wantPendingWorkloadsStatus           *kueue.ClusterQueuePendingWorkloadsStatus
+	}{
+		"taking snapshot of cluster queue is disabled": {},
+		"taking snapshot of cluster queue is enabled": {
+			queueVisibilityClusterQueuesMaxCount: 2,
+			queueVisibilityUpdateInterval:        10 * time.Millisecond,
+			wantPendingWorkloadsStatus: &kueue.ClusterQueuePendingWorkloadsStatus{
+				Head: []kueue.ClusterQueuePendingWorkload{
+					{Name: "one"}, {Name: "two"},
+				},
+			},
+		},
+		"verify the head of pending workloads when the number of pending workloads exceeds MaxCount": {
+			queueVisibilityClusterQueuesMaxCount: 1,
+			queueVisibilityUpdateInterval:        10 * time.Millisecond,
+			wantPendingWorkloadsStatus: &kueue.ClusterQueuePendingWorkloadsStatus{
+				Head: []kueue.ClusterQueuePendingWorkload{
+					{Name: "one"},
+				},
+			},
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			cq := utiltesting.MakeClusterQueue(cqName).
+				QueueingStrategy(kueue.StrictFIFO).Obj()
+			lq := utiltesting.MakeLocalQueue(lqName, "").
+				ClusterQueue(cqName).Obj()
+			ctx := context.Background()
+
+			cl := utiltesting.NewClientBuilder().WithLists(defaultWls).WithObjects(lq, cq).WithStatusSubresource(lq, cq).
+				Build()
+			cCache := cache.New(cl)
+			qManager := queue.NewManager(cl, cCache)
+			if err := qManager.AddClusterQueue(ctx, cq); err != nil {
+				t.Fatalf("Inserting clusterQueue in manager: %v", err)
+			}
+			if err := qManager.AddLocalQueue(ctx, lq); err != nil {
+				t.Fatalf("Inserting localQueue in manager: %v", err)
+			}
+
+			r := NewClusterQueueReconciler(
+				cl,
+				qManager,
+				cCache,
+				WithQueueVisibilityUpdateInterval(tc.queueVisibilityUpdateInterval),
+				WithQueueVisibilityClusterQueuesMaxCount(tc.queueVisibilityClusterQueuesMaxCount),
+			)
+
+			if err := startSnapshotWithTimeout(ctx, r); err != nil {
+				t.Errorf("error startSnapshotWithTimeout: %v", err)
+			}
+
+			if diff := cmp.Diff(tc.wantPendingWorkloadsStatus, r.getWorkloadsStatus(cq), cmpopts.IgnoreFields(kueue.ClusterQueuePendingWorkloadsStatus{}, "LastChangeTime")); len(diff) != 0 {
+				t.Errorf("unexpected ClusterQueueStatus (-want,+got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func startSnapshotWithTimeout(ctx context.Context, r *ClusterQueueReconciler) error {
+	ctx, cancel := context.WithTimeout(ctx, snapshotTimeout)
+	defer cancel()
```

Comment:
IIUC the `defer` should be called outside at the end of test case. Now it would be called when leaving `startSnapshotWithTimeout` which might be too early

Reply:
Actually, the cancel function is not of much interest to us here, the main idea to sent a signal "Done" to "Start" function when will "snapshotTimeout" come. It's most similar to "time.Sleep".

Reply:
I'm afraid we may have a race condition, that if the cancel happens before the snapshot is computed (and technically it could as the code is async), then we will not get the status update.

Reply:
corrected https://github.com/kubernetes-sigs/kueue/pull/1069/commits/9e80a72b116bbded09adce1f1164e1577c2002ca

Reply:
I'm still suspicious of this, see https://github.com/kubernetes-sigs/kueue/pull/1069#discussion_r1318730569.